core/chains/evm/logpoller/orm.go

```diff
@@ -386,25 +393,96 @@ type Exp struct {
 }
 
 func (o *DSORM) SelectUnmatchedLogIDs(ctx context.Context, limit int64) (ids []uint64, err error) {
-	query := `
-		SELECT l.id FROM evm.logs l LEFT JOIN (
+	batchLogsSubQuery := `SELECT id, evm_chain_id, address, event_sig FROM evm.logs
+                                WHERE evm_chain_id = $1 AND block_number >= $2 AND block_number <= $3`
+
+	query := fmt.Sprintf(`
+		SELECT l.id FROM (%s) l LEFT JOIN (
 			SELECT evm_chain_id, address, event
 			FROM evm.log_poller_filters
 				WHERE evm_chain_id = $1
 				GROUP BY evm_chain_id, address, event
 		) r ON l.evm_chain_id = r.evm_chain_id AND l.address = r.address AND l.event_sig = r.event
 		WHERE l.evm_chain_id = $1 AND r.evm_chain_id IS NULL
-	`
+	`, batchLogsSubQuery)
 
-	if limit == 0 {
-		err = o.ds.SelectContext(ctx, &ids, query, ubig.New(o.chainID))
+	latestBlock, err := o.SelectLatestBlock(ctx)
+	if err != nil {
 		return ids, err
 	}
-	err = o.ds.SelectContext(ctx, &ids, fmt.Sprintf("%s LIMIT %d", query, limit), ubig.New(o.chainID))
+
+	_, err = o.ExecPagedQuery(ctx, limit, latestBlock.FinalizedBlockNumber, func(lower, upper int64) (numResults int64, err2 error) {
+		var rowIDs []uint64
+		err2 = o.ds.SelectContext(ctx, &rowIDs, query, ubig.New(o.chainID), lower, upper)
+		// dedupe rowIDs before appending them to results
+		m := make(map[uint64]bool)
+		for _, id := range rowIDs {
+			if _, val := m[id]; !val {
+				m[id] = true
+				ids = append(ids, id)
+			}
+		}
+		return int64(len(rowIDs)), err2
+	})
 
 	return ids, err
 }
 
+// SelectExcessLogIDs finds any logs old enough that MaxLogsKept has been exceeded for every filter they match.
+func (o *DSORM) SelectExcessLogIDs(ctx context.Context, limit int64) (results []uint64, err error) {
+	withSubQuery := ` -- Roll up the filter table into 1 row per filter
+		SELECT name,
+				ARRAY_AGG(address) AS addresses, ARRAY_AGG(event) AS events,
+				(ARRAY_AGG(topic2) FILTER(WHERE topic2 IS NOT NULL)) AS topic2,
```

Comment:
Are these topics used for any filtering at this moment? Maybe we should not pollute the query by aggregating these topics if we don't need them, wdyt?

Reply:
Not at the moment, but I've been trying to include them in any new queries we add so that it will be less work when we eventually merge the PR that implements them. I don't think they affect the query time noticeably.

Reply:
Not at the moment, but I've been trying to include them in any new queries we add so that it will be less work when we eventually merge the PR that implements them. I don't think they affect the query time noticeably.

Reply:
It's more of a readability thing. If I go to code and see these options used in the query I might incorrectly assume this feature is fully supported and should just work. If you plan to implement that soon we can keep it, but if we don't have any exact plans about supporting that feature then we can drop them. Kinda YAGNI rule. Up to you ;)

Reply:
Alright